File: src/cli/src/journal.rs
//! 日志(journal)读写:collect 记录 + 条目解析。
//!
//! 布局:<workdir>/journal/YYYY-MM-DD.md,条目格式:
//! ```markdown
//! ## 09:30
//! 想法文本
//! ```

use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};

use crate::frontmatter::{self, FrontMatter};

/// 一条日志条目。
#[derive(Debug, Clone, PartialEq)]
pub struct Entry {
    /// 来源文件名,如 `2026-08-15.md`
    pub file: String,
    /// 条目 id,时间戳连字符形式,如 `09-30`
    pub id: String,
    /// 条目标题,如 `09:30`
    pub title: String,
    /// 条目正文
    pub text: String,
}

impl Entry {
    /// 引用形式:`2026-08-15.md#09-30`
    pub fn reference(&self) -> String {
        format!("{}#{}", self.file, self.id)
    }
}

/// 一个日志文件:文件名 + front matter + 条目列表。
#[derive(Debug, Clone, PartialEq)]
pub struct JournalFile {
    pub file: String,
    pub fm: FrontMatter,
    pub entries: Vec<Entry>,
}

pub fn journal_dir(workdir: &Path) -> PathBuf {
    workdir.join("journal")
}

/// C-Capture:把一条想法追加到今日日志(自动 ## HH:MM 条目)。
pub fn collect(workdir: &Path, text: &str) -> Result<PathBuf, String> {
    let dir = journal_dir(workdir);
    fs::create_dir_all(&dir).map_err(|e| format!("mkdir journal: {e}"))?;
    let now = chrono::Local::now();
    let path = dir.join(format!("{}.md", now.format("%Y-%m-%d")));
    let title = now.format("%H:%M").to_string();

    let mut content = String::new();
    let mut existing_ids: Vec<String> = vec![];
    if path.exists() {
        let prev = fs::read_to_string(&path).map_err(|e| format!("read: {e}"))?;
        existing_ids = read_entries(&prev).into_iter().map(|e| e.id).collect();
        content.push_str(&prev);
        if !content.ends_with('\n') {
            content.push('\n');
        }
        content.push('\n');
    }
    // 标题带序号(同分钟多条):`## 19:53`、`## 19:53-2`、`## 19:53-3`…
    let base = now.format("%H-%M").to_string();
    let same = existing_ids
        .iter()
        .filter(|id| **id == base || id.starts_with(&format!("{}-", base)))
        .count();
    let title = if same == 0 {
        title
    } else {
        format!("{}-{}", title, same + 1)
    };
    content.push_str(&format!("## {}\n{}\n", title, text.trim()));
    fs::write(&path, content).map_err(|e| format!("write: {e}"))?;
    Ok(path)
}

/// 解析日志正文为条目列表(识别 `## ` 小节)。
/// id 直接由标题生成:`## 19:53` → `19-53`;`## 19:53-2` → `19-53-2`
/// (同分钟多条时 collect 会在标题中写序号,故 id 稳定、删除条目不漂移)。
pub fn read_entries(body: &str) -> Vec<Entry> {
    let mut out: Vec<Entry> = vec![];
    let mut current: Option<(String, String, Vec<String>)> = None; // (title, id, lines)
    for line in body.lines() {
        if let Some(title) = line.strip_prefix("## ") {
            if let Some((t, id, lines)) = current.take() {
                out.push(Entry {
                    file: String::new(), // 由调用方填充
                    id,
                    title: t,
                    text: lines.join("\n").trim().to_string(),
                });
            }
            let title = title.trim().to_string();
            let id = title.replace(':', "-");
            current = Some((title, id, vec![]));
        } else if let Some((_, _, lines)) = &mut current {
            lines.push(line.to_string());
        }
    }
    if let Some((t, id, lines)) = current.take() {
        out.push(Entry {
            file: String::new(),
            id,
            title: t,
            text: lines.join("\n").trim().to_string(),
        });
    }
    out
}

/// 读取全部日志文件(按文件名排序),解析 front matter 与条目。
pub fn read_all(workdir: &Path) -> Result<Vec<JournalFile>, String> {
    let dir = journal_dir(workdir);
    let mut out = vec![];
    if !dir.exists() {
        return Ok(out);
    }
    for entry in fs::read_dir(&dir).map_err(|e| format!("read_dir journal: {e}"))? {
        let path = entry.map_err(|e| format!("read_dir entry: {e}"))?.path();
        if path.extension().map(|e| e == "md").unwrap_or(false) {
            let content = fs::read_to_string(&path).map_err(|e| format!("read {}: {e}", path.display()))?;
            let (fm, body) = frontmatter::parse(&content);
            let file = path.file_name().unwrap().to_string_lossy().to_string();
            let mut entries = read_entries(&body);
            for e in &mut entries {
                e.file = file.clone();
            }
            out.push(JournalFile {
                file,
                fm: fm.unwrap_or_default(),
                entries,
            });
        }
    }
    out.sort_by(|a, b| a.file.cmp(&b.file));
    Ok(out)
}

/// 更新单个日志文件的 front matter:读 → 闭包修改 → 写回。
/// 文件原本无 front matter 时,若闭包产生了非空 topics/sources 则创建。
pub fn update_front_matter(
    workdir: &Path,
    file: &str,
    f: impl FnOnce(&mut FrontMatter),
) -> Result<(), String> {
    let path = journal_dir(workdir).join(file);
    let content = fs::read_to_string(&path).map_err(|e| format!("read {}: {e}", path.display()))?;
    let (fm, body) = frontmatter::parse(&content);
    let mut fm = fm.unwrap_or_default();
    f(&mut fm);
    let new_doc = format!("{}{}", frontmatter::render(&fm), body);
    fs::write(&path, new_doc).map_err(|e| format!("write {}: {e}", path.display()))?;
    Ok(())
}

/// 各日志条目 id → 主题(合并所有文件)。
pub fn all_topic_assignments(journals: &[JournalFile]) -> BTreeMap<String, String> {
    let mut out = BTreeMap::new();
    for jf in journals {
        for (k, v) in &jf.fm.topics {
            out.insert(format!("{}#{}", jf.file, k), v.clone());
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_entries() {
        let body = "## 09:30\n宣传册逻辑\n\n## 12:15\n每一篇独立发布\n";
        let entries = read_entries(body);
        assert_eq!(entries.len(), 2);
        assert_eq!(entries[0].id, "09-30");
        assert_eq!(entries[0].text, "宣传册逻辑");
        assert_eq!(entries[1].title, "12:15");
        assert_eq!(entries[1].reference(), "#12-15");
    }

    #[test]
    fn parse_entries_with_sequence_ids() {
        let body = "## 19:53\n第一条\n\n## 19:53-2\n第二条\n\n## 19:53-3\n第三条\n";
        let entries = read_entries(body);
        assert_eq!(entries.len(), 3);
        assert_eq!(entries[0].id, "19-53");
        assert_eq!(entries[1].id, "19-53-2");
        assert_eq!(entries[2].id, "19-53-3");
    }

    #[test]
    fn collect_creates_and_appends() {
        let dir = std::env::temp_dir().join(format!("material-test-{}", std::process::id()));
        let _ = fs::remove_dir_all(&dir);
        let path = collect(&dir, "第一条想法").unwrap();
        assert!(path.exists());
        collect(&dir, "第二条想法").unwrap();
        let content = fs::read_to_string(&path).unwrap();
        assert!(content.contains("第一条想法"));
        assert!(content.contains("第二条想法"));
        let entries = read_entries(&content);
        assert_eq!(entries.len(), 2);
        let _ = fs::remove_dir_all(&dir);
    }
}

File: src/studio/lib/workflow/journal.dart
/// 日志（journal）读写：collect 记录 + 条目解析（与 CLI journal.rs 对齐）。
///
/// 布局：workdir/journal/YYYY-MM-DD.md，条目格式：
/// ```markdown
/// ## 09:30
/// 想法文本
/// ```
library;

import 'dart:io';

import 'package:path/path.dart' as p;

import 'frontmatter.dart';

/// 一条日志条目。
class Entry {
  /// 来源文件名，如 `2026-08-15.md`
  final String file;
  /// 条目 id，时间戳连字符形式，如 `09-30`
  final String id;
  /// 条目标题，如 `09:30`
  final String title;
  /// 条目正文
  final String text;

  const Entry({
    required this.file,
    required this.id,
    required this.title,
    required this.text,
  });

  /// 引用形式：`2026-08-15.md#09-30`
  String reference() => '$file#$id';
}

/// 一个日志文件：文件名 + front matter + 条目列表。
class JournalFile {
  final String file;
  final FrontMatter fm;
  final List<Entry> entries;

  const JournalFile({
    required this.file,
    required this.fm,
    required this.entries,
  });
}

String journalDir(String workdir) => p.join(workdir, 'journal');

/// C-Capture：把一条想法追加到今日日志（自动 ## HH:MM 条目）。
Future<String> collect(String workdir, String text, {DateTime? now}) async {
  final t = now ?? DateTime.now();
  final title = '${t.hour.toString().padLeft(2, '0')}:${t.minute.toString().padLeft(2, '0')}';
  return appendEntry(workdir, title, text, now: t);
}

/// 追加一条条目到今日日志；同标题多条时标题追加序号（`## 09:30-2`）。
Future<String> appendEntry(
  String workdir,
  String title,
  String text, {
  DateTime? now,
}) async {
  final t = now ?? DateTime.now();
  final dir = journalDir(workdir);
  Directory(dir).createSync(recursive: true);
  final dateStr = '${t.year.toString().padLeft(4, '0')}-'
      '${t.month.toString().padLeft(2, '0')}-'
      '${t.day.toString().padLeft(2, '0')}';
  final path = p.join(dir, '$dateStr.md');

  final buffer = StringBuffer();
  final existingIds = <String>[];
  final file = File(path);
  if (file.existsSync()) {
    final prev = file.readAsStringSync();
    existingIds.addAll(readEntries(prev).map((e) => e.id));
    buffer.write(prev);
    if (!prev.endsWith('\n')) buffer.write('\n');
    buffer.write('\n');
  }
  final baseId = title.replaceAll(':', '-');
  final same = existingIds
      .where((id) => id == baseId || id.startsWith('$baseId-'))
      .length;
  final finalTitle = same == 0 ? title : '$title-${same + 1}';
  buffer.write('## $finalTitle\n${text.trim()}\n');
  file.writeAsStringSync(buffer.toString());
  return path;
}

/// 解析日志正文为条目列表（识别 `## ` 小节）。
/// id 直接由标题生成：`## 19:53` → `19-53`；`## 19:53-2` → `19-53-2`。
List<Entry> readEntries(String body) {
  final out = <Entry>[];
  String? curTitle;
  String? curId;
  final lines = <String>[];

  void flush() {
    if (curTitle != null) {
      out.add(Entry(
        file: '',
        id: curId!,
        title: curTitle,
        text: lines.join('\n').trim(),
      ));
    }
    lines.clear();
  }

  for (final line in body.split('\n')) {
    if (line.startsWith('## ')) {
      flush();
      curTitle = line.substring(3).trim();
      curId = curTitle.replaceAll(':', '-');
    } else if (curTitle != null) {
      lines.add(line);
    }
  }
  flush();
  return out;
}

/// 读取全部日志文件（按文件名排序），解析 front matter 与条目。
Future<List<JournalFile>> readAll(String workdir) async {
  final dir = Directory(journalDir(workdir));
  final out = <JournalFile>[];
  if (!dir.existsSync()) return out;
  final files = dir
      .listSync()
      .whereType<File>()
      .where((f) => p.extension(f.path) == '.md')
      .toList()
    ..sort((a, b) => p.basename(a.path).compareTo(p.basename(b.path)));
  for (final f in files) {
    final content = f.readAsStringSync();
    final (fm, body) = FrontMatter.parse(content);
    final file = p.basename(f.path);
    final entries = readEntries(body).map((e) => Entry(
          file: file,
          id: e.id,
          title: e.title,
          text: e.text,
        )).toList();
    out.add(JournalFile(file: file, fm: fm ?? FrontMatter(), entries: entries));
  }
  return out;
}

/// 更新单个日志文件的 front matter：读 → 闭包修改 → 写回。
Future<void> updateFrontMatter(
  String workdir,
  String file,
  void Function(FrontMatter fm) update,
) async {
  final path = p.join(journalDir(workdir), file);
  final content = File(path).readAsStringSync();
  final (fm, body) = FrontMatter.parse(content);
  final f = fm ?? FrontMatter();
  update(f);
  File(path).writeAsStringSync('${f.render()}$body');
}

/// 收集归属 `topic` 的全部日志条目（按文件 + 时间排序）。
/// 无归属条目时返回空列表。
Future<List<Entry>> collectForTopic(String workdir, String topic) async {
  final journals = await readAll(workdir);
  final collected = <Entry>[];
  for (final jf in journals) {
    for (final entry in jf.fm.topics.entries) {
      if (entry.value == topic) {
        final e = jf.entries.where((e) => e.id == entry.key).firstOrNull;
        if (e != null) collected.add(e);
      }
    }
  }
  collected.sort((a, b) {
    final byFile = a.file.compareTo(b.file);
    return byFile != 0 ? byFile : a.id.compareTo(b.id);
  });
  return collected;
}

/// 各日志条目 id → 主题（合并所有文件）。
Map<String, String> allTopicAssignments(List<JournalFile> journals) {
  final out = <String, String>{};
  for (final jf in journals) {
    for (final entry in jf.fm.topics.entries) {
      out['${jf.file}#${entry.key}'] = entry.value;
    }
  }
  return out;
}
